llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -1591,10 +1591,150 @@ static void hoistLockstepIdenticalDbgVariableRecords(
   }
 }
 
+// See if we can intersect the attributes for two callbases (used for
+// hoisting/sinking).
+static std::optional<AttributeList> tryIntersectAttrs(const CallBase *CB0,
+                                                      const CallBase *CB1) {
```

Comment:
This intersection logic should really live in the Attributes implementation, not SimplifyCFG.

Reply:
Thats what I originally had done, the reason I changed it was the sink/hoist intersection isn't necessarily complete as there may be attributes that change the nature of a function to a degree that makes sinking/hoisting no profitable i.e you can imagine something like the alloc attributes which are probably best not to drop.

Reply:
I don't think we are interested in modelling that, but if we really wanted, we could pass a profitability callback.

Reply:
Likewise all the `opt*` attrs. I'll move to attributes but Ill give the API a whitelist of attrs may intersect to avoid these cases.

Reply:
It's a bit borderline, but I'd not consider something like optnone a droppable attribute. Though in any case, these are not relevant to call-sites.

Reply:
Agreed these aren't really relevant to a call-site, but there is nothing correctness related to dropping `optnone` is there? So AFAICT the only argument against `optnone` being dropable is related to "would we want to?" which is based on the usage, hence having the code in simplifycfg where we can actually answer that question. But like you said, its not too difficult to have add a profitability check in the API.

Reply:
PR at https://github.com/llvm/llvm-project/pull/109719